File: java/src/main/java/heritage/Enseignant.java
package heritage;

public class Enseignant extends Personne {
    private String matiere;
    private double salaire;

    public Enseignant(String nom, int age, String matiere, double salaire) {
        super(nom, age);
        setMatiere(matiere);
        setSalaire(salaire);
    }

    public String getMatiere() {
        return matiere;
    }

    public void setMatiere(String matiere) {
        if (matiere == null || matiere.isBlank()) {
            throw new IllegalArgumentException("La matiere est obligatoire");
        }
        this.matiere = matiere;
    }

    public double getSalaire() {
        return salaire;
    }

    public void setSalaire(double salaire) {
        if (salaire < 0.0) {
            throw new IllegalArgumentException("Le salaire doit etre positif");
        }
        this.salaire = salaire;
    }

    @Override
    public String afficherDetails() {
        return "Enseignant{" +
            "nom='" + getNom() + '\'' +
            ", age=" + getAge() +
            ", matiere='" + matiere + '\'' +
            ", salaire=" + salaire +
            '}';
    }

    @Override
    public String toString() {
        return afficherDetails();
    }
}

File: python/src/enseignant.py
from math import isfinite

from src.personne import Personne


class Enseignant(Personne):
    def __init__(self, nom: str, age: int, matiere: str, salaire: float):
        super().__init__(nom, age)
        self.matiere = matiere
        self.salaire = salaire

    @property
    def matiere(self) -> str:
        return self.__matiere

    @matiere.setter
    def matiere(self, value: str) -> None:
        if not isinstance(value, str) or not value.strip():
            raise ValueError("La matiere est obligatoire")
        self.__matiere = value.strip()

    @property
    def salaire(self) -> float:
        return self.__salaire

    @salaire.setter
    def salaire(self, value: float) -> None:
        if not isinstance(value, (int, float)) or isinstance(value, bool) or not isfinite(float(value)):
            raise TypeError("Le salaire doit etre un nombre")
        if value < 0:
            raise ValueError(f"Le salaire doit etre positif ou nul, valeur recue: {value}")
        self.__salaire = float(value)

    def afficher_details(self) -> str:
        return (
            f"Enseignant: {self.nom}, age: {self.age}, "
            f"matiere: {self.matiere}, "
            f"salaire: {self.salaire:.2f}"
        )

    def __repr__(self) -> str:
        return (
            "Enseignant("
            f"nom='{self.nom}', "
            f"age={self.age}, "
            f"matiere='{self.matiere}', "
            f"salaire={self.salaire}"
            ")"
        )
